cranelift/codegen/src/isa/x64/abi.rs

```diff
@@ -29,6 +29,130 @@ pub(crate) type X64Caller = Caller<X64ABIMachineSpec>;
+        //
+        // We generate the "guard area" register which is essentially the framze_size aligned to
+        // guard_size. We copy the stack pointer and and subtract the guard area from it. This
+        // gets us a register that we can use to compare when looping.
+        //
+        // After that we emit the loop, Essentially we just adjust the stack pointer one guard_size'd
+        // distance at a time and then touch the stack by writing anything to it. We use the previously
+        // created "guard area" register to know when to stop looping.
+        //
+        // When we have touched all the pages that we need, we have to restore the stack pointer
+        // to where it was before.
+        //
+        // If you are editing this code, make sure to manually update the jump offset below
+        // We don't have relocations/labels on this part of the pipeline, so we need
+        // to manually do the offsets.
+        //
+        // Generate the following code:
+        //         mov  tmp_reg, rsp
+        //         sub  tmp_reg, GUARD_SIZE * probe_count
+        // .loop:
+        //         sub  rsp, GUARD_SIZE
+        //         mov  [rsp], 0
+        //         cmp  rsp, tmp_reg
+        //         jne  .loop
+        //         add  rsp, GUARD_SIZE * probe_count
+
+        // Create the guard bound register
+        // mov  tmp_reg, rsp
+        insts.push(Inst::gen_move(
+            Writable::from_reg(tmp_reg),
+            regs::rsp(),
+            I64,
+        ));
+        // sub  tmp_reg, GUARD_SIZE * probe_count
+        insts.push(Inst::alu_rmi_r(
+            OperandSize::Size64,
+            AluRmiROpcode::Sub,
+            RegMemImm::imm(guard_size * probe_count),
+            Writable::from_reg(tmp_reg),
+        ));
+
+        // Emit the main loop!
+        // sub  rsp, GUARD_SIZE
+        insts.push(Inst::alu_rmi_r(
+            OperandSize::Size64,
+            AluRmiROpcode::Sub,
+            RegMemImm::imm(guard_size),
+            Writable::from_reg(regs::rsp()),
+        ));
+
+        // TODO: `mov [rsp], 0` would be better, but we don't have that instruction
+        // Probe the stack! We don't use Inst::gen_store_stack here because we need a predictable
+        // instruction size.
+        // mov  [rsp], rsp
+        insts.push(Inst::mov_r_m(
+            OperandSize::Size32, // Use Size32 since it saves us one byte
+            regs::rsp(),
+            SyntheticAmode::Real(Amode::imm_reg(0, regs::rsp())),
+        ));
+
+        // Compare and jump if we are not done yet
+        // cmp  rsp, tmp_reg
+        insts.push(Inst::cmp_rmi_r(
+            OperandSize::Size64,
+            RegMemImm::reg(regs::rsp()),
+            tmp_reg,
+        ));
+
+        // jne  .loop_start
+        let loop_size = 7 + // sub
+            3 + // mov
+            3 + // cmp
+            2; // jmp
+        insts.push(Inst::jmp_if_rel_offset(CC::NZ, -loop_size));
```

Comment:
I'm not too happy about this, and did this mostly as a proof of concept to see if it fixes the issues we were having in https://github.com/bjorn3/rustc_codegen_cranelift/issues/1261#issuecomment-1219307630 .

I think the real solution here is allowing us to emit labels and use those with the regular jump instruction.

`gen_prologue` curently does two things:
* Initializes part of the `Callee` struct (calculates stack sizes)
* Generates the prologue instructions

These instructions aren't emitted immediatley, instead we call this early on in vcode emission to initialize the `Callee` struct, and the only later on emit the previously generated instructions.

https://github.com/bytecodealliance/wasmtime/blob/d620705a323e3da59bd90473b4e627c8502b1255/cranelift/codegen/src/machinst/vcode.rs#L790-L793

I think we can change this slightly to allow label emission in `gen_prologue` by splitting it into two functions.

One that precomputes the stack sizes and initializes the `Callee` struct, that is called early on (where the current `gen_prologue` currently is).

And a second one that directly emits the prologue instructions. This way we can emit them and at the same time pass a function to generate a label at the current emit point.


Are there issues with emitting labels so late in the pipeline? And does this sound like a reasonable refactor?